### DISCUSSION TITLE
Reset my server. Have a full drive backup. How do I restore everything?

### DISCUSSION THREAD

**Discussion by SangeloDev** [outline/outline#3550] [Self Hosting] [answered]
So, maybe this isn't asked a lot. But how do I import my data to a new Outline installation after I have reset my server, but made a full backup of the hard drive?
I reinstalled my server to have YunoHost, so I can't really reuse my Outline installation again because I think it uses some special configurations and I don't want to dip my toes into that any further because I have managed to break multiple programs in the past by doing so.
That's why I came here to ask. I'd really appreciate any help!

Solution provided by @Limezy:
https://github.com/outline/outline/discussions/3550#discussioncomment-2798220

**Comment by Limezy** [comment, marked as answer]
Hi ! I'm glad you use the Yunohost version ! 

* Did you do an export of your collections before or not ? 
* Did you have a MinIO or Amazon S3 storage setup ? 
* Is the history of your documents important to you ?
* Did you have a small number of users ?
* Was your Outline using the last version when you reset your server ?

I think depending on these parameters I could try to advise an action plan to recover your Outline.

**Reply by SangeloDev** [reply]
> Hi ! I'm glad you use the Yunohost version !
> 
>     * Did you do an export of your collections before or not ?
> 
>     * Did you have a MinIO or Amazon S3 storage setup ?
> 
>     * Is the history of your documents important to you ?
> 
>     * Did you have a small number of users ?
> 
>     * Was your Outline using the last version when you reset your server ?
> 
> 
> I think depending on these parameters I could try to advise an action plan to recover your Outline.

1. No I didn't, I only discovered the option once I reinstalled everything... silly me haha
2. Yes I did
3. Currently, yes, so I would be happy to have them restored. It isn't much though, so if it is completely necessary I could live with it
4. I was the only one using that installation
5. The version was 0.62.0, so the last version, not the latest.

Thanks for taking some time to help me! Really appreciate it.

**Reply by tommoor** [reply]
Really you just need to copy across the Postgres database and the environment variables - everything should "just work" the

**Reply by SangeloDev** [reply]
> Really you just need to copy across the Postgres database and the environment variables - everything should "just work" the 

Can you tell me where the database is located? I searched through the file structure earlier but couldn't seem to find it.

**Reply by tommoor** [reply]
I have no idea where your database is located I'm afraid, it's completely configurable.

**Reply by Limezy** [reply]
OK, so now it's sure you can't have your Outline back "the easy way", by importing some export you did.
Good news is that you should recover all, including history etc... 
Bad news is that the process will probably be a bit painful if you don't have a lot of knowledge of how you did in the first place

You have to give us **much** more information if you want some help : how did you install your first instance ? Following a tutorial on internet ? Using Docker as per the docs ? Building from source ? Do you have somewhere a copy of the .env file ? (If yes please copy it here after removing sensitive information first) What auth provider were you using ? What software or provider were you using for your S3 storage ? Are you willing to use the built-in MinIO from Yunohost ?

I bet you have been using docker. If that's indeed the case, then you'll basically need to find where your postgres image was, extract the database from that image data, re-inject it in the Yunohost postgres database, make a few tweaks in the database to have your Yunohost user match with your previous user. Similar story for your S3 storage : find where the data was, extract it, re-inject it in the Yunohost MinIO... 

Let's start discussing here and in the end, for more Yunohost related stuff we may transit the discussion to Yunohost forums

**Reply by SangeloDev** [reply]
> I have no idea where your database is located I'm afraid, it's completely configurable.

Ah alright, so I was searching in the wrong place, I misunderstood. I'll have to dig deeper inside my backup, and I'll make sure to update you all.

> You have to give us **much** more information if you want some help : how did you install your first instance ? Following a tutorial on internet ? Using Docker as per the docs ? Building from source ? Do you have somewhere a copy of the .env file ? (If yes please copy it here after removing sensitive information first) What auth provider were you using ? What software or provider were you using for your S3 storage ? Are you willing to use the built-in MinIO from Yunohost ?
> 
> I bet you have been using docker. If that's indeed the case, then you'll basically need to find where your postgres image was, extract the database from that image data, re-inject it in the Yunohost postgres database, make a few tweaks in the database to have your Yunohost user match with your previous user. Similar story for your S3 storage : find where the data was, extract it, re-inject it in the Yunohost MinIO...

I built everything from source, as I do remember building everything using ``yarn``. I used the documentation here on the repo to install everything, as I think YouTube tutorials as mostly outdated and I do want to learn new things (also, I didn't find any that were relevant). As I said, I have a full drive backup, so the .env file is safe and sound. I was struggling a bit with authentication, but went with Slack Workspaces in the end. MinIO was configured on my server for storage after reading through that documentation as well. I installed Outline on YunoHost with all recommended dependencies in the pre-install, so I have everything set up there. I have also tested it for a bit, and it seems that it is working fine.
I used docker for a bunch of stuff, not here though, which is why I reset my server, so I have everything organized a bit better. I'll search for my database file and if I find it, I'll update you all here.

**Reply by SangeloDev** [reply]
[Here](https://github.com/outline/outline/files/8716357/dotenv.txt) is the censored .env file that I used.

**Reply by Limezy** [reply]
Ok well noted. I'll have a look at your .env. But what seems very strange to me is that if you built from source, it means you had to setup a working postgres database + redis + minio *yourself*, so you should know approximately how and where you did it don't you ?
Which OS did you build it on ?

**Reply by SangeloDev** [reply]
I know where most of that is, just don't know where PostgreSQL is stored exactly because it came preinstalled on my VPS. Maybe it's the default.
I used Ubuntu 20.04 LTS

**Reply by Limezy** [reply]
Apparently for Ubuntu you should look at `/var/lib/postgresql/10/main` (Maybe the 10 will change here)

**Reply by SangeloDev** [reply]
I used the ``pg_dump`` command as follows: ``pg_dump databasename > databasename.pgsql`` inside a virtual machine with the backup mounted to export the database. Now I would do almost the same to import it into YunoHost: ``pg_dump yunohostdbname < databasename.pgsql`` after uploading it using ftp.
Though I am concerned as to if this could break some things like authentication. Would that be a problem or is it stored seperately?

**Reply by almereyda** [reply]
👏🏽 

In the [README of Outline for Yunohost](https://github.com/YunoHost-Apps/outline_ynh/) it reads, that this is deployed alongisde the [Yunohost Dex App](https://github.com/YunoHost-apps/dex_ynh) for Yunohost.

If you wanted to restore authentication credentials for your Yunohost instance, you would also have to check the Backup and Restore instructions for that module. If what you wanted to achieve is, that you want to be able to start off fresh from a new set of authentication credentials, it seems you're already set and good to go.

**Reply by Limezy** [reply]
Well done ! You now have secured your text and history, which is probably the most important.
Solving Auth will be rather easy, and that's why I would first try to secure the MinIO / images part.
If MinIO was installed on your first setup, you should go to the MinIO data path and copy all that was inside your folder.
You can then go to `/home/yunohost.app/minio/outlinestorage/` (the MinIO bucket is called outlinestorage) on your Yunohost and copy all the content. Then run `chown -R minio:minio *` command, and then `sudo yunohost service restart minio`. 
You can then go to your Yunohost minio storage URL, log in to the console and check that you can look at your different files

**Reply by SangeloDev** [reply]
Done! Seems to work well! Now the last part is the database, and to see if anything breaks.
EDIT: screenshot:
![image](https://user-images.githubusercontent.com/47925793/169569812-15490952-0886-4d2b-bafd-020c3bac7add.png)

**Reply by Limezy** [reply]
Congrats ! You can now import back your Postgres database backup in Yunohost.
To avoid any corrupt data, before and after you should run `sudo yunohost service stop outline` and `sudo yunohost service restart outline`
At this point, you should be able to run Outline without any problem, and have all your data + documents back, using your former auth method (don't forget to reset all cookies before)
Once you get there, the last step will be to manipulate your database so that all the activity and history of your former user will be transferred to your new Yunohost user. To be honest I don't exactly know how to do that... But I'm sure we will find a solution !

@tommoor is there some doc somewhere where I can check how to modify the default auth provider + user ?

**Reply by SangeloDev** [reply]
Kinda stuck here though... Is there a way to enable the login as ``postgres``?

**Reply by Limezy** [reply]
I see. You can first run the command `sudo cat /var/www/outline/.env` and find the line which looks like : 
`DATABASE_URL=postgres://outline:1592DB091801C8BECA355BFC6AB707D987AC51D6DE974295C68E0F5E7055973E@localhost:5432/outline`. Save the `1592DB091801C8BECA355BFC6AB707D987AC51D6DE974295C68E0F5E7055973E` chain somewhere as this is the password of your outline database.

The command you have to run to get into psql interactive mode is : `sudo --login --user=postgres PGUSER=outline PGPASSWORD=1592DB091801C8BECA355BFC6AB707D987AC51D6DE974295C68E0F5E7055973E psql outline`.

I'm trying to check for you what should be the exact command to run `pg_dump` executable

**Reply by Limezy** [reply]
If you want to "see" what you are doing, I would suggest to install the Yunohost adminer app, and use these settings with the password as described above : 
<img width="346" alt="Screenshot 2022-05-21 at 11 06 14" src="https://user-images.githubusercontent.com/24638389/169634767-a16ee587-3cd6-4b3b-94cf-7735381aef51.png">

**Reply by Limezy** [reply]
OK I am trying to migrate an Outline instance and actually is not exactly that simple.
I'll come back here with the list of commands once I've confirmed a "safe" routine to do it !

**Reply by Limezy** [reply]
> Really you just need to copy across the Postgres database and the environment variables - everything should "just work" the

Hi @tommoor after checking, this doesn't really work. I get that error : 
```

Error: error:06065064:digital envelope routines:EVP_DecryptFinal_ex:bad decrypt
    at Decipheriv.final (internal/crypto/cipher.js:174:29)
    at Object.get (/var/www/outline/node_modules/sequelize-encrypted/index.js:41:79)
    at getEncryptedColumn (/var/www/outline/build/server/models/decorators/Encrypted.js:30:60)
    at Object.get jwtSecret (/var/www/outline/build/server/models/User.js:287:46)
    at Object.get (/var/www/outline/node_modules/sequelize/lib/model.js:3498:41)
    at Object.funcs.<computed> (/var/www/outline/node_modules/sequelize/lib/model.js:1105:25)
    at Object.getJwtToken (/var/www/outline/build/server/models/User.js:220:15)
    at signIn (/var/www/outline/build/server/utils/authentication.js:100:41)
    at /var/www/outline/build/server/middlewares/passport.js:65:40
    at callback (/var/www/outline/node_modules/@outlinewiki/koa-passport/lib/framework/koa.js:93:25)
    at Strategy.strategy.success (/var/www/outline/node_modules/passport/lib/middleware/authenticate.js:219:18)
    at verified (/var/www/outline/node_modules/passport-oauth2/lib/strategy.js:189:20)
    at Strategy._verify (/var/www/outline/build/server/routes/auth/providers/slack.js:82:14)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  library: 'digital envelope routines',
  function: 'EVP_DecryptFinal_ex',
  reason: 'bad decrypt',
  code: 'ERR_OSSL_EVP_BAD_DECRYPT'
```
And "Error Looks like something broke!" shows in the UI.
Any idea how to solve this ?

**Reply by tommoor** [reply]
`SECRET_KEY` must match between installations to decrypt some database columns

**Reply by Limezy** [reply]
Obviously 🙄...
Sorry for bothering and thanks for your quick help !

**Reply by tommoor** [reply]
It'll be self explanatory next time someone runs into this…

https://github.com/outline/outline/commit/f2be756cf403f7a3d574d35a3bdec78dc80e3f04#diff-850ef88e92bf5eb6cc5ded9f6b87b815e7a706c2c103dae117e00efde203fc4dR24

**Reply by Limezy** [reply]
OK @SangeloDev here is a routine to fully restore your Outline wiki. I hope I didn't miss anything.
1. Install adminer app in your Yunohost
2. `sudo yunohost service stop outline`
3. `sudo --login --user=postgres dropdb outline`
4. `sudo --login --user=postgres createdb outline`
5. `sudo --login --user=postgres psql outline < your_dump.pgsql` (replace your_dump.pgsql with your own dump file)
6. `cd /var/www/outline` and then `yarn db:migrate` (this is because your database was from a v0.62.0 instance, and Yunohost is v0.63.0)
7. `nano /var/www/outline/.env` and change both SECRET_KEY and UTILS_SECRET using the values from your backup .env
8. From this same file, copy the Postgresql password as you'll need it to login using Adminer app
9. Go to Adminer app and login using the saved password from above
10. Go to `authentication_providers` table, then click on `Select data`
11. Click on the edit pencil button, and change "name" to `oidc` and "providerID" to your main Yunohost domain (for example `mydomain.fr`)
12. `sudo yunohost service restart outline`
13. Remove your web browser cookies for Outline URL, and then login using the standard Yunohost OIDC mechanism (through Dex), using the Yunohost user you want to have replacing your former Outline user
14. At this step you can quickly check if everything seems to be OK (collections, documents etc...)
15. Go to Adminer app again, then `user_authentications` table, then `Select data`
16. You should see two rows, and one of them having "scopes" column showing `{openid,profile,email}`
17. Download both `accessToken` and `refreshToken` files from that row on your computer, and keep somewhere the value of "providerId" from that same row
18. Now let's reset again the database, to remove the "ghost" user we have just created
19. `sudo yunohost service stop outline`
20. `sudo --login --user=postgres dropdb outline`
21. `sudo --login --user=postgres createdb outline`
22. `sudo --login --user=postgres psql outline < your_dump.pgsql` (replace your_dump.pgsql with your own dump file)
23. `cd /var/www/outline` and then `yarn db:migrate` (this is because your database was from a v0.62.0 instance, and Yunohost is v0.63.0)
24. Go to Adminer app and login using the saved password from above
25. Go to `authentication_providers` table, then click on `Select data`
26. Click on the edit pencil button, and change "name" to `oidc` and "providerID" to your main Yunohost domain (for example `mydomain.fr`)
27. Go to `user_authentications` table, then `Select data`
28. Upload both `accessToken` and `refreshToken` files from your computer, change "scopes" to `{openid,profile,email}`, and "providerId" using the value you got at step 17
29. `sudo yunohost service restart outline`
30. You should now be able to have a fully working Outline, with all documents and history, using the Yunohost userbase, and with your Yunohost user mapped to the former user you had in your first setup !

**Reply by SangeloDev** [reply]
Everything worked fine until I came to step 29.
![image](https://user-images.githubusercontent.com/47925793/169688533-d1f5ec8b-0414-402c-8986-40ab5f654f04.png)
![image](https://user-images.githubusercontent.com/47925793/169688577-02d72b67-ca0a-4108-8c4e-b5df5fd09fcf.png)
Did I do something wrong?

**Reply by Limezy** [reply]
Did you redo twice the database drop and creation back ?
It's precisely meant to avoid these double users.
Let me check on my side if I forgot one step

**Reply by SangeloDev** [reply]
I dropped the database twice, yes.

**Reply by Limezy** [reply]
You didn't do anything wrong, there is a small mistake in my tutorial... 
I'm trying to find what I did before writting it. It's something at step 28. Maybe the Id to put back or providerID.
I'll come back once I found it !

**Reply by Limezy** [reply]
I got it ! At step 28, you need to : 
* Upload access and refresh tokens
* set "scopes" as `{openid,profile,email}`
* Set "providerId" using the one you got from step 17

I have edited my steps above accordingly.
If you have it already, you can do steps 19 to 30. If not, you should redo steps 2 to 30
Tell me if it works !

**Reply by SangeloDev** [reply]
This time, I cannot even sign in
![image](https://user-images.githubusercontent.com/47925793/169691019-e6c97c3b-008e-4240-a821-69430da57a96.png)

**Reply by Limezy** [reply]
Did it go through the Dex auth or not even ?
Did you remove all cookies and/or try to use incognito mode ?

**Reply by SangeloDev** [reply]
Yes, it does direct me to Dex, and I have tried clearing cookies and going through incognito. No difference at all.

**Reply by Limezy** [reply]
Can you try again while having a `tail -f /var/log/outline/outline.log` running, and then copy here the related error ? I hope we can get additional information. 
Is login working for another Yunohost user through Dex ? (You can create a dummy one just to try?)

**Reply by SangeloDev** [reply]
```{"error":"User authentication CgdzYW5nZWxvEgRsZGFw already exists for 308f6a58-32bb-4ccd-9c1b-9dc88eae9a86, tried to assign to 0a1a88ad-a9ea-4c37-965f-105a6c818919","stack":"Error: User authentication CgdzYW5nZWxvEgRsZGFw already exists for 308f6a58-32bb-4ccd-9c1b-9dc88eae9a86, tried to assign to 0a1a88ad-a9ea-4c37-965f-105a6c818919\n    at userCreator (/var/www/outline/build/server/commands/userCreator.js:47:13)\n    at runMicrotasks (<anonymous>)\n    at processTicksAndRejections (internal/process/task_queues.js:95:5)\n    at async accountProvisioner (/var/www/outline/build/server/commands/accountProvisioner.js:59:20)\n    at async OAuth2Strategy._verify (/var/www/outline/build/server/routes/auth/providers/oidc.js:93:22)","level":"error","message":"Error during authentication"}```
Looks like it cannot log me in, I'll try with a dummy account.
EDIT: A dummy account logs in fine. Seems like it tries to add something to the database I already added.

**Reply by Limezy** [reply]
Could you screenshot here both your user_authentications and authentification_providers tables ?

**Reply by SangeloDev** [reply]
nvm, I have found the issue. I miswrote my main domain in the authentification_providers table, and now it works well! Thanks for the help!

**Reply by Limezy** [reply]
Yeah, we made it 👍

**Reply by Limezy** [reply]
Well, after review I found that my tutorial above was still missing a few steps. Not necessary to have your instance back, but needed if you want to be able to update your Yunohost Outline version without any hassle !
31. Bring PSQL in interactive mode using `sudo --login --user=postgres psql`
32. Change the outline postgres user password using this command : `ALTER USER outline WITH PASSWORD 'SECRET_KEY';` (replace SECRET_KEY value by the one you got as SECRET_KEY in your backup .env)
33. Then you can quit PSQL interactive mode using `\q` (not mandatory but you can check that the password was changed using Adminer app)
34. Change secret_key value from your Yunohost app instance using the command `sudo yunohost app setting outline secret_key -v SECRET_KEY` (replace SECRET_KEY value by the one you got as SECRET_KEY in your backup .env)
35. Change utils_secret value from your Yunohost app instance using the command `sudo yunohost app setting outline utils_secret -v UTILS_SECRET` (replace UTILS_SECRET value by the one you got as UTILS_SECRET in your backup .env)
36. The next time you'll update Outline from Yunohost you should have a small warning about the fact that your .env has been manually modified but everything should work fine (if not please [report here](forum.yunohost.org/t/outline-wiki-an-open-extensible-wiki-for-your-team-built-using-react-and-node-js/17197/))

**Reply by SangeloDev** [reply]
Having a few problems with that, though maybe unrelated to Outline.
I always get this error: ``Sorry, user admin is not allowed to execute '/bin/bash -c psql' as postgres on hostname.``
Any recommendations?

**Reply by Limezy** [reply]
Can you try to run first `sudo su` ?

**Reply by Limezy** [reply]
And obviously there is step 37 :
37. Change your .env file postgres line as `DATABASE_URL=postgres://outline:SECRET_KEY@localhost:5432/outline`

**Reply by SangeloDev** [reply]
I don't know if I should continue here, though I have a problem with psql not starting now. Should I open a new issue elsewhere?

**Reply by Limezy** [reply]
Could you please start a thread in the Yunohost forum ? We could get help from more experienced users and devs
You should add a copy of the Postgres logs

**Comment by Limezy** [comment]
By the way @tommoor there is obviously no obligation, but now that the Yunohost package is quite stable, I believe you could add it to your documentation as a (non officially supported) installation option, probably best suited for beginners that need a very small instance and don't have the knowledge to properly configure or manage it. I guess that would remove some of the "I can't get images working, please help" or "why do I need Slack to login" stuff happening here in the discussions
